[docs] Adds button styling faq (#12345)
expo/expo: docs/pages/faq/react-native-styling-buttons.md

---
title: 'Styling a React Native Button'
---

> React Native exports a [`<Button />`](https://reactnative.dev/docs/button) component that exposes the native button element for Android, iOS, and the web. The `<Button />` component accepts `title` and `onPress` props, however it does not accept a `style` prop, which makes it hard to customize the style. The closest we can get to styling a `<Button />` exported from React Native is with the `color` prop. Below is an example of two buttons on Android, iOS, and the web. The first button is the default `<Button />` and the second is another default `<Button />` with its `color` prop set to `"red".

Reviewer: ```suggestion
React Native exports a [`<Button />`](https://reactnative.dev/docs/button) component that exposes the native button element for Android, iOS, and the web. The `<Button />` component accepts `title` and `onPress` props but it does not accept a `style` prop, which makes it hard to customize the style. The closest we can get to styling a `<Button />` exported from React Native is with the `color` prop. Below is an example of two buttons on Android, iOS, and the web. The first button is the default `<Button />` and the second is another default `<Button />` with its `color` prop set to `"red".
```

> ![default-button](https://user-images.githubusercontent.com/6455018/112760134-f82c9c80-8fc3-11eb-86fd-59a4d9abf68b.png)

Reviewer: Add image to the repo

> To create a button with a custom style, we can to turn to the [`<Pressable />`](https://reactnative.dev/docs/pressable) component.`<Pressable />`let's us fully customize the appearance of a pressable element (like a button), in addition to allowing us to customize it's behavior. Here's an example of using`<Pressable />` to create a button component:

Reviewer: ```suggestion
To create a button with a custom style, we can to turn to the [`<Pressable />`](../versions/latest/react-native/pressable/) component.`<Pressable />`let's us fully customize the appearance of a pressable element (like a button), in addition to allowing us to customize it's behavior. Here's an example of using`<Pressable />` to create a button component:
```

```jsx
import React from 'react';
import { Text, View, StyleSheet, Pressable } from 'react-native';

> export default function Button(props) {
>   const { onPress, title = 'Save' } = props;
>   return (
>     <Pressable style={styles.button} onPress={onPress}>
>       <Text style={styles.text}>{title}</Text>
>     </Pressable>
>   );
> }

Reviewer 2: ```suggestion
const Button = ({ onPress, title = 'Save' }) => (
  <Pressable style={styles.button} onPress={onPress}>
    <Text style={styles.text}>{title}</Text>
  </Pressable>
);
export default Button;
```
What do you think about this format?

Author: That format works as well. I personally find the keyword `function` nice to read. I also usually name the argument `props` and destructure it below, so that if this were to be typed with TypeScript, we'd get this syntax (which feels readable to me):

```tsx
type Props = {
  prop1: string;
  prop2: number;
  prop3: boolean;
}

function Button (props: Props) {
  const { prop1, prop2, prop3 } = props;
  // ...
}
``` 

The pattern I personally find slightly harder to read quickly is this:

```tsx
const Button =  ({ prop1, prop2, prop3 }: { prop1: string, prop2: number, prop3: boolean }) => {
  // ...
}
```

That said, either is fine. Is there a pattern in docs? If arrow functions with destructured args are the norm, we should change this to be consistent!

Reviewer: Agree with @jonsamp about using the function keyword. It's clearer than `const X = () => ...`. Also, with object literals, `{ fn() {} }` is often clearer than `{ fn: () => {} }`. We intentionally don't configure our linter to require arrow functions wherever possible. Arrow functions make the most sense in callbacks where it's more important to close over the environment.

const styles = StyleSheet.create({
  button: {
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: 12,
    paddingHorizontal: 32,
    borderRadius: 4,
    elevation: 3,
    backgroundColor: 'black',
  },
  text: {
    fontSize: 16,
    lineHeight: 21,
    fontWeight: 'bold',
    letterSpacing: 0.25,
    color: 'white',
  },
});
```

And here's the result of this code:

> <img width="973" alt="Screen Shot 2021-03-28 at 1 02 59 PM" src="/static/images/faq-button-style-pressable.png" />

Reviewer: Give this good alt text

> <img width="973" alt="Screen Shot 2021-03-28 at 1 02 59 PM" src="/static/images/faq-button-style-pressable.png" alt="Custom styled button component using Pressable" />

Reviewer: delete the noisy alt text

> <img width="973" alt="Screen Shot 2021-03-28 at 1 02 59 PM" src="https://user-images.githubusercontent.com/6455018/112760791-34f99300-8fc6-11eb-9acf-459641cb02e2.png" />

Reviewer: Commit this file to the repository and give it useful alt text. Also be sure to crush it with at least `optipng -o7`.

> React Native's `<Button />` component does not accept a `style` prop, and it's `color` prop is limited and appears differently across Android, iOS, and the web. With the `<Pressable />` component, we can create custom buttons that fit our app's design. Those styles will also be the same across Android, iOS, and the web, which will give our apps a consistent look on every platform.

Reviewer: ```suggestion
React Native's `<Button />` component does not accept a `style` prop, and its `color` prop is limited and appears differently across Android, iOS, and the web. With the `<Pressable />` component, we can create custom buttons that fit our app's design. Those styles will also be the same across Android, iOS, and the web, which will give our apps a consistent look on every platform.
```
